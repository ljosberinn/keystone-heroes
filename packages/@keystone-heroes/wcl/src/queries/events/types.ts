export type BaseEvent<T extends Record<string, unknown>> = T & {
  timestamp: number;
  // pin: "0" but irrelevant
};

type EncounterStartEvent = BaseEvent<{
  type: "encounterstart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
  level: number;
  affixes: number[];
}>;

type CombatantInfoEvent = BaseEvent<{
  type: "combatantinfo";
  sourceID: number;
  gear: [];
  auras: {
    source: number;
    ability: number;
    stacks: number;
    icon: string;
    name?: string;
  }[];
  expansion: string;
  faction: number;
  specID: number;
  covenantID: number;
  soulbindID: number;
  strength: number;
  agility: number;
  stamina: number;
  intellect: number;
  dodge: number;
  parry: number;
  block: number;
  armor: number;
  critMelee: number;
  critRanged: number;
  critSpell: number;
  speed: number;
  leech: number;
  hasteMelee: number;
  hasteRanged: number;
  hasteSpell: number;
  avoidance: number;
  mastery: number;
  versatilityDamageDone: number;
  versatilityHealingDone: number;
  versatilityDamageReduction: number;
  talents: { id: number; icon: string }[];
  pvpTalents: { id: number; icon: string }[];
  artifact: { traitID: number; rank: number; spellID: number; icon: string }[];
  heartOfAzeroth: {
    traitID: number;
    rank: number;
    spellID: number;
    icon: string;
  }[];
}>;

export type ApplyDebuffEvent = BaseEvent<{
  type: "applydebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  targetInstance?: number;
  targetMarker?: number;
}>;

export type ApplyDebuffStackEvent = BaseEvent<{
  type: "applydebuffstack";
  stack: number;
  abilityGameID: number;
}>;

export type ApplyBuffEvent = BaseEvent<{
  type: "applybuff";
  sourceID: number;
  sourceInstance?: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
}>;

export type RemoveBuffEvent = BaseEvent<{
  type: "removebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type CastEvent = BaseEvent<{
  type: "cast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker?: number;
}>;

export type DamageEvent = BaseEvent<{
  type: "damage";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  mitigated?: number;
  unmitigatedAmount?: number;
  tick?: boolean;
  sourceMarker?: number;
  absorbed?: number;
  overkill?: number;
  buffs?: string;
}>;

export type BeginCastEvent = BaseEvent<{
  type: "begincast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

export type RemoveDebuffEvent = BaseEvent<{
  type: "removedebuff";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type SummonEvent = BaseEvent<{
  type: "summon";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type PhaseStartEvent = BaseEvent<{
  type: "phasestart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
}>;

export type HealEvent = BaseEvent<{
  type: "heal";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  sourceMarker?: number;
  targetMarker?: number;
  tick?: boolean;
  overheal?: number;
  absorbed?: number;
}>;

type EnergizeEvent = BaseEvent<{
  type: "energize";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  resourceChange: number;
  resourceChangeType: number;
  otherResourceChange: number;
  waste: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type ApplyBuffStackEvent = BaseEvent<{
  type: "applybuffstack";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  stack: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type InterruptEvent = BaseEvent<{
  type: "interrupt";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  sourceMarker?: number;
}>;

export type AbsorbEvent = BaseEvent<{
  type: "absorbed";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  attackerID: number;
  amount: number;
  extraAbilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type RefreshDebuffEvent = BaseEvent<{
  type: "refreshdebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

type DispelEvent = BaseEvent<{
  type: "dispel";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  isBuff: boolean;
  sourceMarker: number;
  targetMarker: number;
}>;

export type DeathEvent = BaseEvent<{
  type: "death";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  // only present if NPC kills player
  killerID?: number;
  killerInstance?: number;
  killingAbilityGameID?: number;
  targetMarker?: number;
}>;

export type AnyEvent =
  | DeathEvent
  | DispelEvent
  | RefreshDebuffEvent
  | AbsorbEvent
  | InterruptEvent
  | ApplyBuffStackEvent
  | EnergizeEvent
  | HealEvent
  | PhaseStartEvent
  | SummonEvent
  | RemoveDebuffEvent
  | BeginCastEvent
  | DamageEvent
  | CastEvent
  | RemoveBuffEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
  | ApplyDebuffEvent
  | CombatantInfoEvent
  | EncounterStartEvent;

export type AllTrackedEventTypes = (
  | CastEvent
  | DeathEvent
  | AbsorbEvent
  | DamageEvent
  | HealEvent
  | InterruptEvent
  | BeginCastEvent
  | ApplyBuffEvent
  | ApplyDebuffEvent
  | ApplyDebuffStackEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[];
