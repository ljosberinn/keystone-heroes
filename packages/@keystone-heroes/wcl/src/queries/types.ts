import type { PlayableClass, Role, SpecName } from "@prisma/client";

export type Table = {
  totalTime: number;
  itemLevel: number;
  logVersion: number;
  gameVersion: number;
  composition: Composition[];
  damageDone: DamageDone[];
  healingDone: HealingDone[];
  damageTaken: DamageTaken[];
  deathEvents: DeathEvent[];
  playerDetails: PlayerDetails;
};

type Composition = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  specs: SpecMeta[];
};

type SpecMeta = {
  spec: SpecName;
  role: Role;
};

export type DamageDone = Pick<Composition, "name" | "id" | "guid"> & {
  total: number;
  icon: string;
  type: Composition["type"] | "Pet" | "NPC";
};

export type HealingDone = DamageDone;

export type DamageTaken = Pick<DamageDone, "name" | "guid" | "total"> & {
  abilityIcon: string;
  type: number;
};

type DeathEvent = Pick<DamageDone, "name" | "id" | "guid" | "type" | "icon"> & {
  deathTime: number;
  ability: Omit<DamageTaken, "total">;
};

export type PlayerDetails = {
  tanks: InDepthCharacterInformation[];
  dps: InDepthCharacterInformation[];
  healers: InDepthCharacterInformation[];
};

export type InDepthCharacterInformation = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  server: string;
  icon: string;
  specs: SpecName[];
  minItemLevel: number;
  maxItemLevel: number;
  combatantInfo: CombatantInfo;
};

type CombatantInfo = {
  stats: Stats;
  talents: Talent[];
  gear: (Item | LegendaryItem)[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  artifact?: CustomPowerSet[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  heartOfAzeroth?: SecondaryCustomPowerSet[];
  customPowerSet: CustomPowerSet[];
  secondaryCustomPowerSet: SecondaryCustomPowerSet[];
  tertiaryCustomPowerSet: unknown[];
  specIDs: number[];
  factionID: number;
  covenantID?: number;
  soulbindID?: number;
};

type Stats = {
  Speed: Stat;
  Dodge: Stat;
  Mastery: Stat;
  Stamina: Stat;
  Haste: Stat;
  Leech: Stat;
  Armor: Stat;
  Agility: Stat;
  Crit: Stat;
  "Item Level": Stat;
  Parry: Stat;
  Avoidance: Stat;
  Versatility: Stat;
  Intellect: Stat;
};

type Stat = {
  min: number;
  max: number;
};

export type Talent = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
};

type CustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export enum ItemQuality {
  POOR = 1,
  COMMON = 2,
  RARE = 3,
  SUPERIOR = 4,
  LEGENDARY = 5,
}

export type Item = {
  id: number;
  slot: number;
  quality: ItemQuality;
  icon: string;
  name?: string;
  itemLevel: number;
  bonusIDs?: number[];
  gems?: Gem[];
  permanentEnchant?: number;
  permanentEnchantName?: string;
  onUseEnchant?: number;
  onUseEnchantName?: string;
  effectID?: number;
  effectIcon?: string;
  effectName?: string;
  temporaryEnchant?: number;
  temporaryEnchantName?: string;
};

export type LegendaryItem = Item & {
  quality: ItemQuality.LEGENDARY;
  effectID: number;
  effectIcon: string;
  effectName: string;
};

type Gem = {
  id: number;
  itemLevel: number;
  icon: string;
};

type SecondaryCustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export type Conduit = SecondaryCustomPowerSet;
export type CovenantTrait = CustomPowerSet;
