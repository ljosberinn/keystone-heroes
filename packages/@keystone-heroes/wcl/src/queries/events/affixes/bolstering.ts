import { Affixes } from "@keystone-heroes/db/types";

import type { AllTrackedEventTypes, ApplyBuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const BOLSTERING = 209_859;

/**
 * @see https://www.warcraftlogs.com/reports/mBf1t94GZ2hRWprv#fight=7&type=auras&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applybuff%22%20and%20ability.id%20%3D%20209859%20and%20target.type%20%3D%20%22NPC%22&hostility=1
 *
 * ```gql
 * {
 *   reportData {
 *     report(code: "mBf1t94GZ2hRWprv") {
 *       fights(fightIDs: [7]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5996572, endTime: 7458759, limit: 2000, filterExpression: "type = \"applybuff\" and ability.id = 209859 and target.type = \"NPC\"") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applybuff" and ability.id = ${BOLSTERING} and target.type = "npc"`,
];

const isBolsteringEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: BOLSTERING,
});

type StackDataset = {
  targetID: number;
  targetInstance: number | null;
  stacks: number;
};

const getHighestBolsteringStack = (
  allEvents: ApplyBuffEvent[]
): ApplyBuffEvent[] => {
  const stacksPerNPC = allEvents.reduce<StackDataset[]>(
    (acc, { targetID, targetInstance = null }) => {
      const existingIndex = acc.findIndex(
        (dataset) =>
          dataset.targetInstance === targetInstance &&
          targetID === dataset.targetID
      );

      if (existingIndex > -1) {
        return acc.map((dataset, index) =>
          index === existingIndex
            ? { ...dataset, stacks: dataset.stacks + 1 }
            : dataset
        );
      }

      return [
        ...acc,
        {
          targetID,
          targetInstance,
          stacks: 1,
        },
      ];
    },
    []
  );

  const highestStack = stacksPerNPC.reduce((acc, dataset) =>
    acc?.stacks >= dataset.stacks ? acc : dataset
  );

  const NPCsWithSameFinalStackAmount = stacksPerNPC.filter(
    (dataset) => dataset.stacks === highestStack.stacks
  );

  return allEvents.reduce<(ApplyBuffEvent & { stacks: number })[]>(
    (acc, event) => {
      const dataset = NPCsWithSameFinalStackAmount.find(
        (npc) =>
          npc.targetID === event.targetID &&
          npc.targetInstance === event.targetInstance
      );

      if (!dataset) {
        return acc;
      }

      const previousMatch = acc.find(
        (oldEvent) =>
          oldEvent.targetID === event.targetID &&
          oldEvent.targetInstance === event.targetInstance
      );

      const enhancedEvent = { ...event, stacks: dataset.stacks };

      if (!previousMatch) {
        return [...acc, enhancedEvent];
      }

      // implicitly means it must be a higher stack amount as targetID and
      // targetInstance match
      const isNewer = event.timestamp > previousMatch.timestamp;

      if (isNewer) {
        return [
          ...acc.filter((oldEvent) => oldEvent !== previousMatch),
          enhancedEvent,
        ];
      }

      return acc;
    },
    []
  );
};

export const getBolsteringEvents = (
  allEvents: AllTrackedEventTypes,
  affixSet: Set<Affixes>
): ApplyBuffEvent[] => {
  if (!affixSet.has(Affixes.Bolstering)) {
    return [];
  }

  return getHighestBolsteringStack(allEvents.filter(isBolsteringEvent));
};
