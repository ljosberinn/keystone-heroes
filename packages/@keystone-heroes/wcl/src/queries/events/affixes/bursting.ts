import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const BURSTING = 243_237;

export const getBurstingDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: BURSTING,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
