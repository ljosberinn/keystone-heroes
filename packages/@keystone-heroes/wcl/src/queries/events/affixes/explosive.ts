import type { AnyEvent, DamageEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const EXPLOSIVE = {
  unit: 120_651,
  ability: 240_446,
};

/**
 * @see https://www.warcraftlogs.com/reports/YrA4zyZgQJvHbn32#fight=1&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24target.id%20%3D120651%20AND%20type%20%3D%20%22damage%22%20AND%20overkill%20%3E%200%5E2%24Off%24%23909049%24expression%24target.type%20%3D%20%22player%22%20and%20rawDamage%20%3E%200%20and%20ability.id%20%3D%20240446
 * ```gql
 * {
 *   reportData {
 *     report(code: "YrA4zyZgQJvHbn32") {
 *       fights(fightIDs: [1]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 87034, endTime: 1527765, limit: 2000, filterExpression: "(target.type = \"player\" and rawDamage > 0 and ability.name = \"Explosion\") OR (target.name = \"Explosives\" AND type = \"damage\" AND overkill > 0)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `target.type = "player" and rawDamage > 0 and ability.id = ${EXPLOSIVE.ability}`,
  `target.id = ${EXPLOSIVE.unit} AND type = "damage" AND overkill > 0`,
];

export const isExplosiveDamageEvent = createIsSpecificEvent<DamageEvent>({
  type: "damage",
  abilityGameID: EXPLOSIVE.ability,
});

export const createIsExplosiveDeathEvent =
  (targetID: number) =>
  (event: AnyEvent): event is DamageEvent => {
    return (
      event.type === "damage" &&
      "targetInstance" in event &&
      event.sourceID !== -1 &&
      event.targetID === targetID
    );
  };

export const findExplosiveTargetID = (allEvents: AnyEvent[]): number | null => {
  const dataset = allEvents.reduce<{
    targetID: null | number;
    targetInstance: null | number;
  }>(
    (acc, event) => {
      if (
        // skip damageEvents
        event.type !== "damage" ||
        // ignore event if target is single instance
        !("targetInstance" in event) ||
        // skip empty instances, just TS here
        event.targetInstance === undefined ||
        // check whether the stored instance is lower than this events
        (acc.targetInstance && event.targetInstance < acc.targetInstance)
      ) {
        return acc;
      }

      return {
        targetID: event.targetID,
        targetInstance: event.targetInstance,
      };
    },
    {
      targetID: null,
      targetInstance: null,
    }
  );

  return dataset.targetID;
};
