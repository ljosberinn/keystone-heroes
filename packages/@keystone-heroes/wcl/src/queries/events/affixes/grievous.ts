import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const GRIEVOUS_WOUND = 240_559;

export const getGrievousDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: GRIEVOUS_WOUND,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
