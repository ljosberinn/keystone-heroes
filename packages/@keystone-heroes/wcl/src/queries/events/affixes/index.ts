import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
