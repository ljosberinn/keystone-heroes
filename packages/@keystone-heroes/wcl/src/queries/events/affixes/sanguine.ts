import { EventDataType, HostilityType } from "../../../types";
import type { HealEvent, DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const SANGUINE_ICHOR_HEALING = 226_510;
export const SANGUINE_ICHOR_DAMAGE = 226_512;

export const getSanguineHealingDoneEvents = async (
  params: GetEventBaseParams
): Promise<[HealEvent]> => {
  const allEvents = await getEvents<HealEvent>({
    ...params,
    dataType: EventDataType.Healing,
    hostilityType: HostilityType.Enemies,
    abilityID: SANGUINE_ICHOR_HEALING,
  });

  return [
    allEvents.reduce<HealEvent>((acc, event) => {
      return {
        ...acc,
        amount: acc.amount + event.amount,
        overheal: (acc.overheal ?? 0) + (event.overheal ?? 0),
      };
    }, allEvents[0]),
  ];
};

export const getSanguineDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: SANGUINE_ICHOR_DAMAGE,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
