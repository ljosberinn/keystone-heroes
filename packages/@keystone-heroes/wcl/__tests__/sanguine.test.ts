import { Affixes } from "@keystone-heroes/db/types";

import { getSanguineEvents } from "../src/queries/events/affixes/sanguine";
import allEvents from "./fixtures/allEvents.json";

describe("getSanguineEvents", () => {
  test("works", () => {
    expect(
      getSanguineEvents(allEvents, new Set([Affixes.Sanguine]))
    ).toMatchSnapshot();
  });

  test("does nothing if affix is absent", () => {
    expect(getSanguineEvents(allEvents, new Set())).toHaveLength(0);
  });
});
