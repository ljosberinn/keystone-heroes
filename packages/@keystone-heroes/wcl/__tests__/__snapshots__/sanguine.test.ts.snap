// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSanguineEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 226510,
    "amount": 2265944,
    "hitType": 1,
    "overheal": 83228,
    "pin": "0",
    "sourceID": -1,
    "targetID": 1076,
    "targetInstance": 3,
    "tick": true,
    "timestamp": 12878155,
    "type": "heal",
  },
  Object {
    "abilityGameID": 226512,
    "amount": 10522,
    "hitType": 1,
    "mitigated": 165,
    "pin": "0",
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 294,
    "targetMarker": 2,
    "tick": true,
    "timestamp": 12897350,
    "type": "damage",
    "unmitigatedAmount": 5426,
  },
  Object {
    "abilityGameID": 226512,
    "amount": 46138,
    "hitType": 1,
    "mitigated": 8212,
    "pin": "0",
    "sourceID": -1,
    "sourceMarker": 6,
    "targetID": 330,
    "targetMarker": 6,
    "tick": true,
    "timestamp": 13129280,
    "type": "damage",
    "unmitigatedAmount": 16354,
  },
  Object {
    "abilityGameID": 226512,
    "amount": 6006,
    "hitType": 1,
    "mitigated": 328,
    "pin": "0",
    "sourceID": -1,
    "targetID": 320,
    "tick": true,
    "timestamp": 13553089,
    "type": "damage",
    "unmitigatedAmount": 6334,
  },
  Object {
    "abilityGameID": 226512,
    "absorbed": 5417,
    "amount": 5417,
    "hitType": 3,
    "mitigated": 447,
    "pin": "0",
    "sourceID": -1,
    "targetID": 307,
    "tick": true,
    "timestamp": 14138169,
    "type": "damage",
    "unmitigatedAmount": 5864,
  },
]
`;
